README.md
[![Build Status](https://travis-ci.org/finaldie/skull.svg?branch=master)](https://travis-ci.org/finaldie/skull)
[![GitHub license](https://img.shields.io/github/license/finaldie/skull.svg)]()
[![Platform](https://img.shields.io/badge/platform-Linux-blue.svg)]()
![GitHub release](https://img.shields.io/github/release/finaldie/skull.svg)
![Docker Pulls](https://img.shields.io/docker/pulls/finaldie/skull.svg)
![GitHub closed pull requests](https://img.shields.io/github/issues-pr-closed/finaldie/skull.svg)
![GitHub language count](https://img.shields.io/github/languages/count/finaldie/skull.svg)
![GitHub code size in bytes](https://img.shields.io/github/languages/code-size/finaldie/skull.svg)

Skull
=====
_Skull_ is an event-driven serving framework with multiple modern designs to allow user:
* Fast to create a prototype
* Easy to maintain even a huge project
* Write code in lock-free environment

It helps people to build an application/server extremely fast and strong, high scalability and flexibility in application layer, and with strong performance in engine level. Read more [here][6].

_Skull_ can be used in generic serving layer or embedded device. E.g. _web logic server_, _game server_, etc. Want an example? Check this [_DNSTurbo_][16] project, and watch its [_Trailer on Youtube_][19] for a quick start :)

## Releases
[Changelog](ChangeLog.md)

## Latest Feature Demo
**Memory tracing reports**:
* _Memory Usage_
* _Cross-Scope Malloc/Free_
* _Memory Leak_

![skull-trace](https://github.com/finaldie/misc/blob/master/articals/skull/tracer/skull-trace-newformat.gif)

## Documentations
* [High Level Introduction][6]
* [How To Start][7]
* [How to Test][8]
* [How to Deploy][9]
* [Monitoring][10]
* [API Docs - Cpp][11]
* [API Docs - Python][12]
* [Integrate with Nginx][13]

## How to Build
Use _Ubuntu18.04_ as an example. (Tested on `Debian jessie/stretch`, `Ubuntu 14.04/16.04/18.04`, `alpine 3.6/3.8`, `RHEL6/7` and `Raspbian jessie/stretch`)

### Basic System Requirements
_Skull_ requires:
 - `Platform`: _Linux_
 - `C/C++`: _Compiler supports at least `c99`/`c++11` standards or higher_
 - `Python`: _Python3_

### Install Dependencies
```console
# Install System Dependencies
sudo apt-get install autoconf libtool valgrind expect python3-dev python3-pip;
sudo pip3 install PyYAML pympler WebOb;

# Clone and Build Dependencies (For example: project folder is 'skull')
cd skull && git submodule update --init --recursive;

# Compile and install dependencies
make -j4 dep; sudo make install-dep;
```

### Build and Install
```console
make -j4
sudo make install
```

**Notes:**
 * To disable `jemalloc`, use `make -j4 disable_jemalloc=true` to build it
 * For some _Linux_ Releases, we might need to use `CFLAGS`, `CXXFLAGS`, `LDFLAGS` to finish the build

**More Options:**
 * `python_path`: By default it's python3, but we can override it to another path for testing purpose

### Docker Images
Also, the [**_Docker images_**][31] are ready now, if people don't want to waste time to set up a brand new environment, we can run the **_Docker_** image directly within 1 min :)

Assume that we've already [installed _Docker_][30], then apply the following commands to pull and run the image to bring user into a development ready environment super fast:
```console
-bash$ docker pull finaldie/skull:1.2-build
-bash$ docker run -it finaldie/skull:1.2-build
```

And the below table is the current images:<br>

Tag              | Dockerfile         | Notes                         |
-----------------|--------------------|-------------------------------|
1.4              | ([Dockerfile][20]) | 1.4 Debian runtime image      |
1.4-build        | ([Dockerfile][21]) | 1.4 Debian dev/building image |
1.4-ubuntu       | ([Dockerfile][22]) | 1.4 Ubuntu runtime image      |
1.4-ubuntu-build | ([Dockerfile][23]) | 1.4 Ubuntu dev/building image |
1.4-alpine       | ([Dockerfile][24]) | 1.4 Alpine runtime image      |
1.4-alpine-build | ([Dockerfile][25]) | 1.4 Alpine dev/building image |

## A Quick Demo
After installing _Skull_ into the system, run the following steps to
create a _Skull_ project, have fun :)

### Create a Skull Project
[![skull demo 1](http://g.recordit.co/6yGrVG7i0s.gif)]()

**Notes:**
 * By default, a new module contains the example code which is used to echo-back message
 * Above creation is for `C++` module, type `py` in language selection step if needed

### Playing with Skull
[![skull demo 2](http://g.recordit.co/vSON9N6nuV.gif)]()

## Existing Services
Name                  | Description |
----------------------|-------------|
[DNS Client][2] | Async DNS client for A/AAAA record |
[Http Client][15] | Async http client service, easy to send/receive http request/response |

## Other Resources
Name                  | Description |
----------------------|-------------|
[Skull-Perf Cases][5] | Including some basic perf cases |
[DNSTurbo][16]        | Smart DNS Client based on _Skull_. [Trailer][17] |

## Contribution and Discussion
To discuss any issues, there are some ways we can use:
 - Open an issue on Github directly
 - Discuss at Reddit: [https://www.reddit.com/r/skullengine/][18]
 - Send an email to skull-engine@googlegroups.com
 - Go directly to the [Mail Group][14], and post the questions.

To fix a bug or add a new feature, just **`Fork`** the repo, then apply the fixes/features via a PR.

[1]: https://github.com/finaldie/skull-admin-c
[2]: https://github.com/finaldie/skull-service-dns
[3]: https://developers.google.com/protocol-buffers/
[4]: https://github.com/finaldie/final_libs
[5]: https://github.com/finaldie/skull-perf
[6]: https://github.com/finaldie/skull/wiki
[7]: https://github.com/finaldie/skull/wiki/How-To-Start
[8]: https://github.com/finaldie/skull/wiki/How-To-Test
[9]: https://github.com/finaldie/skull/wiki/How-To-Deploy
[10]: https://github.com/finaldie/skull/wiki/Monitoring
[11]: https://github.com/finaldie/skull/wiki/API-Doc-:-Cpp
[12]: https://github.com/finaldie/skull/wiki/API-Doc-:-Python
[13]: https://github.com/finaldie/skull/wiki/Integrate-with-Nginx
[14]: https://groups.google.com/forum/#!forum/skull-engine
[15]: https://github.com/finaldie/skull-service-httpcli
[16]: https://github.com/finaldie/DNSTurbo
[17]: https://github.com/finaldie/DNSTurbo#trailer
[18]: https://www.reddit.com/r/skullengine/
[19]: https://www.youtube.com/watch?v=2u_-Wl7sDdA

[20]: https://github.com/finaldie/dockerfiles/blob/master/skull/1.4/Dockerfile
[21]: https://github.com/finaldie/dockerfiles/blob/master/skull/1.4/Dockerfile.build
[22]: https://github.com/finaldie/dockerfiles/blob/master/skull/1.4/ubuntu/Dockerfile
[23]: https://github.com/finaldie/dockerfiles/blob/master/skull/1.4/ubuntu/Dockerfile.build
[24]: https://github.com/finaldie/dockerfiles/blob/master/skull/1.4/alpine/Dockerfile
[25]: https://github.com/finaldie/dockerfiles/blob/master/skull/1.4/alpine/Dockerfile.build

[30]: https://docs.docker.com/engine/installation/linux/docker-ce/ubuntu/
[31]: https://hub.docker.com/r/finaldie/skull/


ChangeLog.md
- 2019-01-08 1.5.2
  - Engine
    - **New:** Add building option for minize the dependency
    - **Enhancement:** Static link libyaml to reduce external dependency
    - **Enhancement:** APIs loading can be omitted instead of exit
    - **Enhancement:** Show API loading status in admin `info` command
  - API
    - **Enhancement:** Static link libprotobuf `api-cpp` to reduce dependency
    - **Enhancement:** Refine python APIs
  - MISC
    - **Enhancement:** Add `api-py=false` makefile option
    - **Enhancement:** Refine makefile vars and templates
- 2019-01-02 1.5.1
  - Engine
    - **Fix:** Prevent crash if ep pool run out of file descriptor resource
    - **Fix:** `ep_pool` wrongly calling `create_tcp_entity` in `_create_entity_udp`
  - API
    - **Fix:** Reduce unnecessary syscall during python log
- 2019-01-01 1.4.3
  - API
    - **Enhancement:** Refine C++ APIs to make service and ep APIs more user friendly
    - **Enhancement:** Enhance python suppression file on ubuntu 18
- 2018-12-31 1.4.2
  - Engine
    - **Enhancement:** Add more counters for logger
  - API
    - **Enhancement:** Refine python APIs according to pylint
    - **Enhancement:** Re-structure common folder and makes python common
      source folder align with lib folder structure
    - **Enhancement:** Refine python template module.py according to PEG8
  - CLI
    - **Enhancement:** Refine mem tracing report format
    - **Enhancement:** Add whitelist for memleak report
- 2018-12-21 1.4.1
  - Engine
    - **New:** Introduce tracing level into `sk_malloc`
    - **New:** Add `sk_mon_set` api
    - **Fix:** Service cron job crash issue
    - **Fix:** Refine **Admin Module** `info` and `counter` output
    - **Enhancement:** Add more counters for logger
    - **Enhancement:** Introduce tracing level concept
  - CLI
    - **New:** `skull-trace` dumps stack frames in memleak report
    - **Enhancement:** `skull-trace` refine output and optimize tracing performance
- 2018-12-17 1.3.3
  - Engine
    - **New:** Expose clock status in admin module
    - **Fix:** Replace all `ftime_gettime` with `sk_time_us`
    - **Enhancement:** Make cron timer more accurate
    - **Enhancement:** Add more apis for `sk_time`
  - MISC
    - **Enhancement:** Upgrade `flibs` to v1.2.2
- 2018-12-05 1.3.2
  - Engine
    - **New:** Setup cron job to dump mem stats
    - **Fix:** Fix crash when exiting if tracing enabled
    - **Fix:** Fix mem stat usage for `realloc`
    - **Enhancement:** Switch mem trace log to diag.log instead of stderr
    - **Enhancement:** Refine `sk_malloc` to make it independently
    - **Enhancement:** Split atomic APIs into `sk_atomic`
  - CLI
    - **New:** `skull-trace` generates reports for stats summary, cross-scope and leaking
  - MISC
    - **Fix:** Fix valgrind on alpine-3.8
    - **Enhancement:** Upgrade `protobuf` to v3.6.1
- 2018-08-05 1.3.1
  - Engine
    - **Fix:** Inaccurate memory stat in `ep_send`
  - MISC
    - **Enhancement:** Fix typos
- 2018-07-25 1.2.3
  - Engine
    - **Fix:** Solve memory API re-entrance issue
    - **Enhancement:** Refine _admin_ `memory` command output (more clear)
    - **Enhancement:** Refine _admin_ `info` command output (more clear)
  - CLI
    - **New:** Memory realtime tracing tool `skull trace`
  - MISC
    - **Enhancement:** Cleanup makefile templates
- 2018-07-24 1.2.2
  - Engine
    - **Fix:** Memory stat measurement inaccurate issue in endpoint and service job
  - CLI
    - **New:** New `skull-config` for making makefiles clearly
    - **Enhancement:** Won't create `.skull/makefiles` folder
  - MISC
    - **New:** `memory detail|full` available in _Admin Module_
    - **Enhancement:** Won' copy `.supp` files to `project/bin` folder
- 2018-07-19 1.2.1
  - Engine
    - **Enhancement:** Remove protobuf-c from dependency
  - API
    - **New:** Upgrade python2 to python3
  - CLI
    - **New:** Upgrade python2 to python3
  - MISC
    - **Enhancement:** Move all `_BSD_SOURCE` to `_DEFAULT_SOURCE`
- 2018-06-27 1.1.3
  - Engine
    - **New:** override libc malloc to better measure memory stats
    - **Fix:** txnlog and slowlog can be present in the same transaction
    - **Enhancement:** _Admin Module_ parses commands into an array, more flexible for subcommands
  - MISC
    - **Enhancement:** Upgrade jemalloc to 5.0.1
    - **Enhancement:** Upgrade travis to ubuntu.trusty
- 2017-12-07 1.1.2
  - Engine
    - **New:** Add `Slowlog` option in global config with microsecond precision
  - CLI
    - **Fix:** Fix a unbreak issue in `skull config` command
- 2017-08-23 1.1.1
  - Engine
    - **New:** Docker integrated
    - **New:** Add `command_bind` config item
    - **Fix:** Fix FT Makefile helper message
    - **Fix:** Fix epoll-loop issue when `concurrency=0`
    - **Enhancement:** Remove all the `__WORDSIZE` macro
    - **Enhancement:** Compatible with `musl`
    - **Enhancement:** Reorder Makefile dependency, make build easier
    - **Enhancement:** Compatible with `alpine` Linux release
    - **Enhancement:** Rename `trigger` to `driver`
  - API
    - **Fix:** Fix a unstable UT failure issue in alpine
  - CLI
    - **Fix:** Config name change from concurrent to concurrency
    - **Enhancement:** Refine output message of creating workflow
    - **Enhancement:** Refine output message of creating module
    - **Enhancement:** Refine output message of creating service
    - **Enhancement:** Make CLI more portable
    - **Enhancement:** Add `musl.supp` valgrind suppression file
    - **Enhancement:** For skull-start, to allow not only inside a skull project
    - **Enhancement:** Workflow port bind to 0.0.0.0 by default
    - **Enhancement:**
  - MISC
    - **New:** Support `--no-log-rolling` command arg
    - **New:** Support logs forward to stdout `--std-forwarding`
    - **Fix:** Fix FT unstable issue in docker
    - **Enhancement:** Refine Makefile targets
    - **Enhancement:** Refine config comments
    - **Enhancement:** Upgrade `flibs` to v1.0

src/pto/sk_pto.c
#include <stdlib.h>

#include "api/sk_pto.h"
#include "api/sk_utils.h"

extern sk_proto_ops_t sk_pto_ops_tcp_accept;
extern sk_proto_ops_t sk_pto_ops_workflow_run;
extern sk_proto_ops_t sk_pto_ops_entity_destroy;
extern sk_proto_ops_t sk_pto_ops_srv_iocall;
extern sk_proto_ops_t sk_pto_ops_srv_task_run;
extern sk_proto_ops_t sk_pto_ops_srv_task_done;
extern sk_proto_ops_t sk_pto_ops_timer_triggered;
extern sk_proto_ops_t sk_pto_ops_timer_emit;
extern sk_proto_ops_t sk_pto_ops_std_start;
extern sk_proto_ops_t sk_pto_ops_svc_timer_run;
extern sk_proto_ops_t sk_pto_ops_svc_timer_done;
extern sk_proto_ops_t sk_pto_ops_srv_task_cb;

sk_proto_t sk_pto_tbl[] = {
    {SK_PTO_TCP_ACCEPT,      SK_PTO_PRI_7, 1, 0, 10, &sk_pto_ops_tcp_accept},
    {SK_PTO_WORKFLOW_RUN,    SK_PTO_PRI_5, 0, 0, 11, &sk_pto_ops_workflow_run},
    {SK_PTO_ENTITY_DESTROY,  SK_PTO_PRI_9, 0, 0, 12, &sk_pto_ops_entity_destroy},
    {SK_PTO_SVC_IOCALL,      SK_PTO_PRI_6, 5, 0, 13, &sk_pto_ops_srv_iocall},
    {SK_PTO_SVC_TASK_RUN,    SK_PTO_PRI_4, 5, 0, 14, &sk_pto_ops_srv_task_run},
    {SK_PTO_SVC_TASK_DONE,   SK_PTO_PRI_6, 3, 0, 15, &sk_pto_ops_srv_task_done},
    {SK_PTO_TIMER_TRIGGERED, SK_PTO_PRI_8, 3, 0, 16, &sk_pto_ops_timer_triggered},
    {SK_PTO_TIMER_EMIT,      SK_PTO_PRI_8, 7, 0, 17, &sk_pto_ops_timer_emit},
    {SK_PTO_STD_START,       SK_PTO_PRI_7, 0, 0, 18, &sk_pto_ops_std_start},
    {SK_PTO_SVC_TIMER_RUN,   SK_PTO_PRI_8, 6, 0, 19, &sk_pto_ops_svc_timer_run},
    {SK_PTO_SVC_TIMER_DONE,  SK_PTO_PRI_8, 3, 0, 20, &sk_pto_ops_svc_timer_done},
    {SK_PTO_SVC_TASK_CB,     SK_PTO_PRI_6, 5, 0, 21, &sk_pto_ops_srv_task_cb},
    {SK_PTO_MAX, 0, 0, 0, 0, NULL}
};

#define SK_PTO_BODY(hdr) ((char*)hdr + sizeof(sk_pto_hdr_t))
#define SK_PTO_ARGSZ sizeof(sk_arg_t)
#define SK_PTO_CHKBITS (0xFFFF)

#define checksum(data, sz, salt) (_checksum(data, sz, salt) & SK_PTO_CHKBITS)

static
uint32_t _checksum(void* data, size_t sz, uint32_t salt) {
    if (!data || !sz) return 0;

    uint32_t checksum = 0;

    char const* c = data;
    for (size_t i = 0; i < sz; i++, c++) {
        uint32_t uc = (uint32_t)(*c) ? (uint32_t)(*c) : 1;
        checksum += uc * salt;
        //sk_print("_checksum: %u, salt: %u, sz: %zu, i: %zu, c: %u, uc: %u\n",
        //         checksum, salt, sz, i, *c, uc);
    }

    return checksum;
}

size_t    sk_pto_pack (uint32_t id, sk_pto_hdr_t* hdr, va_list ap) {
    if (!hdr) return 0;

    sk_proto_t* pto = &sk_pto_tbl[id];
    SK_ASSERT_MSG(id == pto->id, "Mis-matched pto id: %u - %u\n", id, pto->id);
    hdr->id = id & 0xFF;
    hdr->argc = pto->argc;

    sk_arg_t* p = (sk_arg_t*)SK_PTO_BODY(hdr);
    int i = 0;
    for (; i < pto->argc; i++, p++) {
        *p = va_arg(ap, sk_arg_t);
    }

    SK_ASSERT_MSG(i == hdr->argc, "Mis-matched argc when create pto: %u\n", id);

    size_t body_sz = SK_PTO_ARGSZ * (size_t)i;
    uint32_t checksum = checksum(SK_PTO_BODY(hdr), body_sz, pto->salt);
    hdr->chk = checksum & SK_PTO_CHKBITS;
    sk_print("sk_pto_pack: id: %u, argc: %u, rawchk: %u, chk: %u\n",
             hdr->id, hdr->argc, checksum, hdr->chk);
    return sizeof(*hdr) + body_sz;
}

sk_arg_t* sk_pto_arg  (uint32_t id, sk_pto_hdr_t* hdr, int idx) {
    if (!hdr || !hdr->argc) return NULL;
    SK_ASSERT_MSG(id < SK_PTO_MAX, "pto id must < pto max: %u\n", id);
    SK_ASSERT_MSG(id == hdr->id, "pto id mis-match: %u - %u\n", id, hdr->id);

    sk_proto_t* pto = &sk_pto_tbl[hdr->id];
    SK_ASSERT_MSG(idx < pto->argc, "Request pto idx must < argc: %d\n", idx);

    return ((sk_arg_t*)SK_PTO_BODY(hdr)) + idx;
}

bool sk_pto_check(uint32_t id, sk_pto_hdr_t* hdr) {
    if (!hdr || !hdr->argc) return true;
    SK_ASSERT_MSG(id < SK_PTO_MAX, "pto id must < pto max: %u\n", id);
    SK_ASSERT_MSG(id == hdr->id, "pto id mis-match: %u - %u\n", id, hdr->id);

    sk_proto_t* pto = &sk_pto_tbl[hdr->id];
    return
        hdr->chk == checksum(SK_PTO_BODY(hdr), SK_PTO_BODYSZ(hdr), pto->salt);
}

